### PR TITLE
feature: support nvidia-container 2.0 to enable GPU access

### LINKED ISSUE #824: [feature request] nvidia GPU native support in runc of Pouch
### Ⅰ. Issue Description

In old days, we use `nvidia-docker` as a plugin to support GPU devices in container. And applications in container use specific cuda library to take advantages of GPU devices. nvidia-docker will get the underlying devices on the host and transfer this into the container.

While now, we could support this on OCI runtime runc by customizing runc like project https://github.com/NVIDIA/nvidia-container-runtime 

As a result, we could make pouch natively support GPU devices without any plugins. In additions, we could make this in Kubernetes cluster to support device plugins.

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1.
2.
3.

### Ⅵ. Environment:

- pouch version (use `pouch version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by CodeJuan**
### Ⅰ. Describe what this PR did
Make pouch natively support GPU devices

### Ⅱ. Does this pull request fix one issue?

fixes #824 #2033 

### Ⅳ. Describe how to verify it
1. Prerequisites: 
  - An instance with a NVIDIA GPU
  - The appropriate GPU driver is installed
2. Create container with `NvidiaConfig`.
```
curl --unix-socket /run/pouchd.sock -X POST http:/containers/create -H "Content-Type: application/json" -d '{"Image": "centos:7", "HostConfig":{"Devices": [],"NvidiaConfig":{"NvidiaDriverCapabilities":"all","NvidiaVisibleDevices":"0"}},"Cmd": ["ping","127.0.0.1"]}'
```
3. Start container.
4. Exec nvidia-smi in container.
5. If nvidia-smi return error `NVIDIA-SMI couldn't find libnvidia-ml.so library in your system. Please make sure that the NVIDIA Display Driver is properly installed and present in your system`, then add a link to libnvidia-ml.so
```
ln -s  /usr/lib64/libnvidia-cfg.so.1 /usr/lib64/libnvidia-cfg.so
ln -s /usr/lib64/libnvidia-cfg.so.387.26 /usr/lib64/libnvidia-cfg.so.1
ln -s  /usr/lib64/libnvidia-ml.so.1 /usr/lib64/libnvidia-ml.so
ln -s /usr/lib64/libnvidia-ml.so.387.26 /usr/lib64/libnvidia-ml.so.1
```

### CODE REVIEW

**Comment by allencloud** [comment]
Why does this pull request not trigger the travisCI? 
@CodeJuan @fuweid

File: daemon/mgr/container_utils.go [hunk 1]
```diff
@@ -379,3 +384,77 @@ func amendContainerSettings(config *types.ContainerConfig, hostConfig *types.Hos
...
 }
+
+// validateNvidiaConfig
+func validateNvidiaConfig(hostConfig *types.HostConfig) error {
+	r := &hostConfig.Resources
+	if r.NvidiaConfig == nil {
+		return nil
+	}
+
+	if err := vlidateNvidiaDriver(r); err != nil {
```
Comment (allencloud) [inline daemon/mgr/container_utils.go]: s/validate/vlidate?

File: daemon/mgr/container_utils.go [hunk 2]
```diff
@@ -379,3 +384,77 @@ func amendContainerSettings(config *types.ContainerConfig, hostConfig *types.Hos
...
+	}
+
+	if err := vlidateNvidiaDevice(hostConfig); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func stringInSlice(a string, list []string) bool {
```
Comment (allencloud) [inline daemon/mgr/container_utils.go]: You do not need to add this function, this has been covered in `utils.StringInSlice`.

File: daemon/mgr/spec_hook.go [hunk 3]
```diff
@@ -152,6 +158,24 @@ func setMountTab(ctx context.Context, c *Container, spec *SpecWrapper) error {
...
+	path, err := exec.LookPath("nvidia-container-runtime-hook")
+	if err != nil {
+		return err
+	}
+	args := []string{path}
+	nvidiaPrestart := specs.Hook{
+		Path: path,
+		Args: append(args, "prestart"),
+	}
+	spec.s.Hooks.Prestart = append(spec.s.Hooks.Prestart, nvidiaPrestart)
```
Comment (allencloud) [inline daemon/mgr/spec_hook.go]: If we already has a prestart hook, will the nvidia prestart take effect here? Just double check this.

**Comment by allencloud** [comment]
How to use this? Do we need to support this in pouch CLI? @CodeJuan

**Comment by CodeJuan** [comment]
@allencloud 
#2033 
I will do it by next Monday.

File: daemon/mgr/spec_hook.go (hunk 3 above)
Comment (CodeJuan) [reply daemon/mgr/spec_hook.go]: Sure, prestart commands are executed after the container namespaces are created.

File: daemon/mgr/container_utils.go [hunk 4]
```diff
@@ -5,14 +5,14 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/opencontainers/selinux/go-selinux/label"
```
Comment (fuweid) [inline daemon/mgr/container_utils.go]: move the third party to next to the pouch group.

File: daemon/mgr/container_validation.go [hunk 5]
```diff
@@ -2,12 +2,16 @@ package mgr
...
+	"strings"
+
+	"github.com/sirupsen/logrus"
 
 	"github.com/alibaba/pouch/apis/types"
 	"github.com/alibaba/pouch/daemon/logger/syslog"
 	"github.com/alibaba/pouch/pkg/system"
-
-	"github.com/sirupsen/logrus"
+	"github.com/alibaba/pouch/pkg/utils"
```
Comment (fuweid) [inline daemon/mgr/container_validation.go]: in pouch, we use three groups like:

```
	"os"
	"strconv"
	"strings"

 	"github.com/alibaba/pouch/apis/types"
	"github.com/alibaba/pouch/daemon/logger/syslog"
	"github.com/alibaba/pouch/pkg/system"
	"github.com/alibaba/pouch/pkg/utils"

 	"github.com/sirupsen/logrus"
```

File: daemon/mgr/container_validation.go [hunk 6]
```diff
@@ -188,3 +196,77 @@ func (mgr *ContainerManager) validateLogConfig(c *Container) error {
...
+		// use default driver capability: utility
+		return nil
+	}
+
+	if n.NvidiaDriverCapabilities == "all" {
+		// enable all capabilities
+		return nil
+	}
+
+	supportedDrivers := []string{"compute", "compat32", "graphics", "utility", "video", "display"}
```
Comment (fuweid) [inline daemon/mgr/container_validation.go]: could we use map global var to hold the supportedDrivers, like

```
var supportedNvidiaDrivers = map[string]struct{}{
        "compute": struct{},
        ....
}
```

therefore, we don't need to allocate new memory in the `utils.StringInSlice`.

File: hack/package/rpm/build.sh [hunk 7]
```diff
@@ -69,6 +74,24 @@ function build_pouch()
     popd
 }
 
+# install nvidia-container-runtime
+function build_nvidia_runtime(){
```
Comment (fuweid) [inline hack/package/rpm/build.sh]: next time, I will rewrite the package folder

**Comment by allencloud** [comment]
Thanks for your great work. I will merge this. @CodeJuan

**Comment by allencloud** [comment]
Do we have a document to help users use GPU with pouchcontainer? @CodeJuan

**Pull request merged**